/*

   calin/proto/diagnostics/simple_charge_hists.proto -- Stephen Fegan -- 2020-04-17

   Protobufs for (simple-sum) channel charge hists

   Copyright 2020, Stephen Fegan <sfegan@llr.in2p3.fr>
   Laboratoire Leprince-Ringuet, CNRS/IN2P3, Ecole Polytechnique, Institut Polytechnique de Paris

   This file is part of "calin"

   "calin" is free software: you can redistribute it and/or modify it
   under the terms of the GNU General Public License version 2 or
   later, as published by the Free Software Foundation.

   "calin" is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

*/

syntax = "proto3";
option cc_enable_arenas = true;

import "calin.proto";
import "math/histogram.proto";

package calin.ix.diagnostics.simple_charge_hists;

message SingleGainSimpleChargeHistsConfig {
  bool enable_hists                                                 = 1 [
    (CFO).desc = "Enable histogramming for this gain channel." ];

  calin.ix.math.histogram.AccumulatedAndSerializedHistogram1DConfig full_wf_qsum = 10 [
    (CFO).desc = "Configuration for full-waveform charge sum." ];

  calin.ix.math.histogram.AccumulatedAndSerializedHistogram1DConfig full_wf_max = 11 [
    (CFO).desc = "Configuration for full-waveform maximum sample." ];

  calin.ix.math.histogram.AccumulatedAndSerializedHistogram1DConfig full_wf_max_index = 12 [
    (CFO).desc = "Configuration for full-waveform index of maximum sample." ];

  calin.ix.math.histogram.AccumulatedAndSerializedHistogram1DConfig opt_win_qsum = 13 [
    (CFO).desc = "Configuration for optimal-window charge sum." ];

  calin.ix.math.histogram.AccumulatedAndSerializedHistogram1DConfig opt_win_qtsum = 14 [
    (CFO).desc = "Configuration for optimal-window amplitude-time sum." ];

  calin.ix.math.histogram.AccumulatedAndSerializedHistogram1DConfig opt_win_index = 15 [
    (CFO).desc = "Configuration for optimal-window index." ];

  calin.ix.math.histogram.AccumulatedAndSerializedHistogram1DConfig ped_win_qsum = 16 [
    (CFO).desc = "Configuration for fixed pedestal-window charge sum." ];

  calin.ix.math.histogram.AccumulatedAndSerializedHistogram1DConfig sig_win_qsum = 17 [
    (CFO).desc = "Configuration for fixed signal-window charge sum." ];
};

message SimpleChargeHistsConfig {
  SingleGainSimpleChargeHistsConfig high_gain                       = 10 [
    (CFO).desc = "Configuration for high or single-gain histograms." ];
  SingleGainSimpleChargeHistsConfig low_gain                        = 20 [
    (CFO).desc = "Configuration for low-gain histograms." ];
};

message OneGainSimpleChargeHists {
  calin.ix.math.histogram.Histogram1DData full_wf_qsum              = 10 [
    (CFO).desc = "Histogram of sum of samples over full waveform.",
    (CFO).units = "DC (bin-units), events (value-units)" ];

  calin.ix.math.histogram.Histogram1DData full_wf_max               = 11 [
    (CFO).desc = "Histogram of maximum sample value over full waveform.",
    (CFO).units = "DC (bin-units), events (value-units)" ];

  calin.ix.math.histogram.Histogram1DData full_wf_max_index         = 12 [
    (CFO).desc = "Histogram of index of maximum sample over full waveform.",
    (CFO).units = "samples (bin-units), events (value-units)" ];

  calin.ix.math.histogram.Histogram1DData opt_win_qsum              = 13 [
    (CFO).desc = "Histogram of sum of samples over optimised window.",
    (CFO).units = "DC (bin-units), events (value-units)" ];

  calin.ix.math.histogram.Histogram1DData opt_win_qtsum             = 14 [
    (CFO).desc = "Histogram of sum of samples * sample index over optimised window.",
    (CFO).units = "DC samples (bin-units), events (value-units)" ];

  calin.ix.math.histogram.Histogram1DData opt_win_index             = 15 [
    (CFO).desc = "Histogram of index of start of optimised window.",
    (CFO).units = "samples (bin-units), events (value-units)" ];

  calin.ix.math.histogram.Histogram1DData ped_win_qsum              = 16 [
    (CFO).desc = "Histogram of sum of samples over fixed pedestal window.",
    (CFO).units = "DC (bin-units), events (value-units)" ];

  calin.ix.math.histogram.Histogram1DData sig_win_qsum              = 17 [
    (CFO).desc = "Histogram of sum of samples over fixed signal window.",
    (CFO).units = "DC (bin-units), events (value-units)" ];
}

message SimpleChargeHists {
  repeated OneGainSimpleChargeHists high_gain_channel               = 1 [
    (CFO).desc = "High gain (or single gain) histograms." ];
  repeated OneGainSimpleChargeHists low_gain_channel                = 2 [
    (CFO).desc = "Low gain (or single gain) histograms." ];
};
